### PR TITLE
feat(certs_tests e2e): add initial e2e certs tests

### LINKED ISSUE #32: write tests for certificates

### PR DESCRIPTION
**Pull request by vdice**
Closes #32

### CODE REVIEW

File: tests/certs_test.go [hunk 1]
```diff
@@ -1,28 +1,476 @@
...
+
+				Entry("a non-wildcard cert to a wildcard domain",
+					domains["wildcard"], certs["www"].Name, certs["www"].CertPath, certs["www"].KeyPath),
+				Entry("a non-wildcard cert to a non-wildcard domain",
+					domains["foo"], certs["foo"].Name, certs["foo"].CertPath, certs["foo"].KeyPath),
+				Entry("a non-wildcard cert to a non-wildcard domain",
+					domains["bar"], certs["bar"].Name, certs["bar"].CertPath, certs["bar"].KeyPath),
+			)
+
+			DescribeTable("2 certs (1 wildcard), 3 domains (1 wildcard)",
```
Comment (vdice) [inline tests/certs_test.go]: The table approach, while it nicely reduces lines of code, breaks down a little in this case where we have a different number of certs vs. domains, as well as the `foo.com` domain case relying on state from the previous `*.foo.com` run... Looking for suggestions here -- keep table-driven or go back to (more verbose but involving no janky `if` logic) inline approach?

File: tests/certs_test.go [hunk 2]
```diff
@@ -1,28 +1,422 @@
...
+			once.Do(func() {
+				testApp = deployApp("example-go")
+			})
+			domain = getRandDomain()
+			certName = strings.Replace(domain, ".", "-", -1)
+			customSSLEndpoint = strings.Replace(testApp.URL, "http", "https", 1)
+		})
+
+		// Basic "Smoke" test verifying SSL endpoints
+		It("can add, attach, list, and remove certs, plus verify SSL endpoint(s)", func() {
```
Comment (vdice) [inline tests/certs_test.go]: Do we believe this "smoke" scenario is still needed since we perform the same actions [here?] (https://github.com/deis/workflow-e2e/pull/75/files#diff-4e69408eafdc533cb4f2550813c5ba3cR331) (though in a slightly more complicated scenario)

File: tests/certs_test.go [hunk 3]
```diff
@@ -1,28 +1,422 @@
...
+			// attempt to remove non-existent cert
+			sess, err = start("deis certs:remove %s", nonExistentCertName)
+			Expect(err).NotTo(HaveOccurred())
+			Eventually(sess.Err).Should(Say("404 Not Found"))
+			Eventually(sess).Should(Exit(1))
+
+			removeCerts([]string{certName})
+		})
+
+		It("can add, attach, list, and remove certs: 2 domains, 1 cert", func() {
```
Comment (vdice) [inline tests/certs_test.go]: Similar to question above, Do we believe this scenario is still needed since we perform similar actions [here?] (https://github.com/deis/workflow-e2e/pull/75/files#diff-4e69408eafdc533cb4f2550813c5ba3cR331) (verifying SSL endpoint on domain w/ cert attached, verifying lack of SSL endpoint on domain(s) w/o cert(s) attached)

File: tests/certs_test.go (hunk 1 above)
Comment (vdice) [inline tests/certs_test.go]: Now that helper methods were added (see top of file), it probably makes sense to change this back from a table-driven approach to the standard/in-line/script style.  Thoughts?

File: tests/tests_suite_test.go [hunk 4]
```diff
@@ -247,7 +253,9 @@ func start(cmdLine string, args ...interface{}) (*Session, error) {
 func startCmd(command Cmd) (*Session, error) {
 	cmd := exec.Command("/bin/sh", "-c", command.CommandLineString)
 	cmd.Env = command.Env
-	io.WriteString(GinkgoWriter, fmt.Sprintf("$ %s\n", command.CommandLineString))
+	if debug {
```
Comment (mboersma) [inline tests/tests_suite_test.go]: This change shouldn't be in here--it undoes the fix in #71.

Comment (vdice) [inline tests/tests_suite_test.go]: Oops! My bad @mboersma ... haha had thought I added the `if debug` earlier and was wondering where it went.  So, apologies for causing the first go-round of the bug (#71) and removing this change now.

File: tests/certs_test.go (hunk 2 above)
Comment (bacongobbler) [inline tests/certs_test.go]: Normally I say that it never hurts to have more tests, though this does look like a duplicate. I'd say let's remove it

File: tests/certs_test.go (hunk 3 above)
Comment (bacongobbler) [inline tests/certs_test.go]: kill it

File: tests/certs_test.go (hunk 1 above)
Comment (bacongobbler) [inline tests/certs_test.go]: I think it's fine as-is if it's testing the use cases we want. If you feel like the other approach is better then by all means go for it, but I live by the mantra of "Ship first, refactor later". :)

**Comment by jchauncey** [comment]
I got failures

```
[Fail] Certs with a deployed app multiple domains and certs 3 certs (no wildcards), 3 domains (1 wildcard) [It] a non-wildcard cert to a wildcard domain
/Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:44

[Fail] Certs with a deployed app multiple domains and certs 3 certs (no wildcards), 3 domains (1 wildcard) [It] a non-wildcard cert to a non-wildcard domain
/Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:44

[Fail] Certs with a deployed app multiple domains and certs 3 certs (no wildcards), 3 domains (1 wildcard) [It] a non-wildcard cert to a non-wildcard domain
/Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:44

[Fail] Certs with a deployed app multiple domains and certs 2 certs (1 wildcard), 3 domains (1 wildcard) [It] a wildcard cert to a non-wildcard domain
/Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:44

[Fail] Certs with a deployed app multiple domains and certs 2 certs (1 wildcard), 3 domains (1
```

**Comment by vdice** [comment]
Hmmm @jchauncey line it fails on is `Eventually(listCerts()).Should(Say("No certs"))` -- could you add what the actual output is vs. expected ("No certs")

**Comment by jchauncey** [comment]
```
Certs
/Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:375
  with a deployed app
  /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:374
    multiple domains and certs
    /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:373
      2 certs (1 wildcard), 3 domains (1 wildcard)
      /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/vendor/github.com/onsi/ginkgo/extensions/table/table.go:96
        a non-wildcard cert to a non-wildcard domain [It]
        /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:46

        No future change is possible.  Bailing out early after 0.000s.
        Got stuck at:
                           Name              | Common Name | SubjectAltName |         Expires         |   Fingerprint   | Domains |  Updated   |  Created
            +--------------------------------+-------------+----------------+-------------------------+-----------------+---------+------------+------------+
              my-custom-495148083-domain-com | *.foo.com   | foo.com        | 21 Jan 2017 (in 1 year) | 8F:8E[...]CD:EB |         | 9 Mar 2016 | 9 Mar 2016
              my-custom-942377268-domain-com | *.foo.com   | foo.com        | 21 Jan 2017 (in 1 year) | 8F:8E[...]CD:EB |         | 9 Mar 2016 | 9 Mar 2016

        Waiting for:
            No certs

        /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:44
```

**Comment by jchauncey** [comment]
```
Certs
/Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:375
  with a deployed app
  /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:374
    multiple domains and certs
    /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:373
      2 certs (1 wildcard), 3 domains (1 wildcard)
      /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/vendor/github.com/onsi/ginkgo/extensions/table/table.go:96
        a wildcard cert to a non-wildcard domain [It]
        /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:46

        No future change is possible.  Bailing out early after 0.000s.
        Got stuck at:
                           Name              | Common Name | SubjectAltName |         Expires         |   Fingerprint   | Domains |  Updated   |  Created
            +--------------------------------+-------------+----------------+-------------------------+-----------------+---------+------------+------------+
              my-custom-495148083-domain-com | *.foo.com   | foo.com        | 21 Jan 2017 (in 1 year) | 8F:8E[...]CD:EB |         | 9 Mar 2016 | 9 Mar 2016
              my-custom-942377268-domain-com | *.foo.com   | foo.com        | 21 Jan 2017 (in 1 year) | 8F:8E[...]CD:EB |         | 9 Mar 2016 | 9 Mar 2016

        Waiting for:
            No certs

        /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:44
```

**Comment by jchauncey** [comment]
```
Certs
/Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:375
  with a deployed app
  /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:374
    multiple domains and certs
    /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:373
      3 certs (no wildcards), 3 domains (1 wildcard)
      /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/vendor/github.com/onsi/ginkgo/extensions/table/table.go:96
        a non-wildcard cert to a non-wildcard domain [It]
        /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:46

        No future change is possible.  Bailing out early after 0.000s.
        Got stuck at:
                           Name              | Common Name | SubjectAltName |         Expires         |   Fingerprint   | Domains |  Updated   |  Created
            +--------------------------------+-------------+----------------+-------------------------+-----------------+---------+------------+------------+
              my-custom-495148083-domain-com | *.foo.com   | foo.com        | 21 Jan 2017 (in 1 year) | 8F:8E[...]CD:EB |         | 9 Mar 2016 | 9 Mar 2016
              my-custom-942377268-domain-com | *.foo.com   | foo.com        | 21 Jan 2017 (in 1 year) | 8F:8E[...]CD:EB |         | 9 Mar 2016 | 9 Mar 2016

        Waiting for:
            No certs

        /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:44
```

**Comment by jchauncey** [comment]
```
Certs
/Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:375
  with a deployed app
  /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:374
    can add, attach, list, and remove certs... improperly [It]
    /Users/jonathanchauncey/go/src/github.com/deis/workflow-e2e/tests/certs_test.go:261

    No future change is possible.  Bailing out early after 0.578s.
    Got stuck at:

    Waiting for:
        404 Not Found
```

**Comment by vdice** [comment]
Thanks, @jchauncey - When the expectation failed in  `can add, attach, list, and remove certs... improperly [It]` (could be an actual app bug), some cleanup didn't get executed (namely, removing test certs).  I've moved the cleanup into an `AfterEach` block, where it should have been all along.  Therefore, less chance for residue affecting subsequent tests if an earlier test bails out early on account of a failure.  Give the latest a go when you have a minute.  Thanks!

**Comment by mboersma** [comment]
This is failing for me:
```
Certs with an app yet to be deployed 
  can add, attach, list, and remove certs
  /Users/matt/Projects/src/github.com/deis/workflow-e2e/tests/certs_test.go:164
$ deis login http://deis.192.168.64.2.xip.io:32656 --username=test-81 --password=asdf1234
Logged in as test-81
$ git init
Initialized empty Git repository in /private/var/folders/82/jhp8dfbx3h36dm85qsbfmm_80000gn/T/deis-workflow-test172175498/.git/
$ deis apps:create test-716274203 
Creating Application... done, created test-716274203
Git remote deis added
remote available at ssh://git@deis.192.168.64.2.xip.io:2222/test-716274203.git
$ deis certs:list
No certs

• Failure in Spec Setup (BeforeEach) [2.055 seconds]
Certs
/Users/matt/Projects/src/github.com/deis/workflow-e2e/tests/certs_test.go:395
  with an app yet to be deployed
  /Users/matt/Projects/src/github.com/deis/workflow-e2e/tests/certs_test.go:165
    can add, attach, list, and remove certs [BeforeEach]
    /Users/matt/Projects/src/github.com/deis/workflow-e2e/tests/certs_test.go:164

    No future change is possible.  Bailing out early after 0.000s.
    Got stuck at:
        ssh://git@deis.192.168.64.2.xip.io:2222/test-716274203.git
        
    Waiting for:
        Git remote deis added

    /Users/matt/Projects/src/github.com/deis/workflow-e2e/tests/certs_test.go:141
```

**Comment by mboersma** [comment]
I'm still getting the dreaded "000" http response failure with `curl` in these tests. The difference seems to be using a specific port number rather than implicitly using 443 or 80.

**Comment by vdice** [comment]
@mboersma does `curl`ing a valid https endpoint work for you if port `443` is explicitly assigned as the port (as opposed to whatever is usually included in your app URL)?

**Comment by mboersma** [comment]
@vdice yes, the issue is that we're converting http:// to https:// but leaving the rest of the URL intact, so the port stays the same. If I hack a string replacement to :443 into the tests, everything's happy. Maybe we can strip off the final port from the `customSSLEndpoint` so `http://test-617295266.192.168.64.2.xip.io:31267` will be end up as `https://test-617295266.192.168.64.2.xip.io`? (Maybe use [`urlx`](https://github.com/goware/urlx) as @arschles had once suggested?)

**Comment by mboersma** [comment]
Working great now in my local kube-solo testing:
```console
Ran 9 of 104 Specs in 169.231 seconds
SUCCESS! -- 9 Passed | 0 Failed | 0 Pending | 95 Skipped PASS | FOCUSED
```
